_posts/2018-12-13-arms-final-report.md

```diff
@@ -0,0 +1,151 @@
+
+# Features of ARMS
+
+In the Fall 2018 semester ARMS was developed with very solid features base and is in a good state to be expanded upon. The core of the application revolves around 'Creators.' Each target format that ARMS will support will have an associated Creator with it. Armature takes in command line arguments, such as ```--sdf``` and ```--drawstuff``` and calls the ```create_file()``` method of the Creator that can handle that request. Each creator follows a Creator interface so ```main()``` does not care how a Creator creates it's associated file. It just knows it can and sends off the request and a dictionary with all the objects parsed from the ARMS TOML file. This allows future file formats to be easily added by following the patter established and also allow existing Creators to be edited without effecting the other Creators.
+
+The main features that ARMS supports are:
+* Spheres
+* Boxes
+* Revolute Joints
+* Ball and Socket Joints
+* Prismatic Joints
+* Non-collision Groups (ODE only)
+* Constants to be used as key words in an ARMS file
+* Macros (groups of shapes/joints that can be repeated)
+
+These features were top priority for the ARCS lab and were delivered within the semester and ARMS will be able to be used to further this research now. Through use bugs and other features needed will be found, but this base line of functionality has proven the usefulness of ARMS.
+
+# Future Work
+
+The following is ideas on how the project could be futher developed.
+
+#### Original Syntax and Parser:
+
+As discussed in the Syntax section of this post, a new syntax and parser that is created specifically for our needs would be a great addition to this project.
+
+#### Continued development of the SDF Creator:
+
+The SDF Creator is a top priority at the time this post was written as it is the format used by Gazebo, which is one of the most popular robotics simulation environments. The SDF creator is probably the best place to start when getting familiar with Armature.
+
+#### ODE struct support:
+
+ODE was initially supported by creating a simulation and displaying it using DrawStuff, a simple graphics library used by ODE to display it's simulations. To use ODE in Dr. Clark's own visualizer. A C++ struct object is needed.
+
+[Example Struct Object](https://github.com/anthonyjclark/ODE-examples/blob/master/ugv/main.cpp)
+
+#### SKEL Format Support:
+The SKEL file format is an XML format based on SDF and describes objects for DART. 
+
+[https://dartsim.github.io/skel_file_format.html](https://dartsim.github.io/skel_file_format.html)
+
+
+Other, more specific things that could be done can be found at the project [Trello Board](https://trello.com/b/xtciB7o8/arms)
+
+
+# Example model and associated ARMS code
+![Picture of example ARMS creation](/assets/2018-12-13-arms-final-report/example_created_shape.png)
```

Comment:
Can you show off the macro feature in this example?

Reply:
Oh yeah great idea!

Reply:
Waiting for testing

Reply:
I'm not sure what you mean?

Reply:
Sorry that was so vauge. I just mean I have the macro version ready but haven't tested it to make sure it actually works so I will be resolving this when I have tested it. I guess I didn't need to mention that here heheh.

Reply:
Done